feat(search): Add opinion embeddings generation and caching

ERosendo opened the pull request:
## Fixes
<!-- What bugs does this fix? Use this syntax to auto-close the issue: -->
<!-- https://docs.github.com/en/issues/tracking-your-work-with-issues/using-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword -->
<!-- E.g.: "Fixes: #XYZ" -->
This PR fixes #4561

## Summary
<!-- What does this fix, how did you fix it, what approach did you take, what gotchas are there in your code or compromises did you make? -->
This PR introduces support for computing and indexing opinion embeddings when `html_with_citations` is updated, along with related refactors, caching logic, and tests.

Embedding generation would be only triggered when the `html_with_citations` field is updated. This decision is based on the following observations:

- Nearly all Opinion records already have this field populated (only 513 records are missing content here, based on a recent query).

- `html_with_citations` has the highest priority in the `with_best_text` method, meaning there’s a very high likelihood this field was used when generating the embeddings currently stored in S3.

- The `html_with_citations` field is actively updated in methods such as `recap_document_into_opinions` and `store_opinion_citations_and_update_parentheticals`.

- Given its coverage, it looks like this field should be considered the standardized source of truth for Opinion content.

### Changes

- New microservice endpoint

  - Adds an endpoint to the microservice list to compute embeddings for given text.

- New Celery task

  - Adds `compute_single_opinion_embeddings` task to generate embeddings for an opinion using the `inception-text` microservice.
  - Includes retry logic for transient errors and caches results for 30 minutes.

- Task chaining refactor

  - Updates task chain so that `compute_single_opinion_embeddings` is prepended before `update_es_document` when `html_with_citations` changes.
  - Ensures embeddings are available before the ES document is updated.

- Cache integration in ES updates

  - Tweaks `update_es_document` to fetch embeddings from cache (if present) when updating an `OpinionDocument`.
  - Injects embeddings into the ES update payload, avoiding recomputation.

- FieldTracker adjustments

  - Updates `es_o_field_tracker` to include `html_with_citations` in tracked fields.


## Deployment

**This PR should:**
<!-- The following labels control the deployment of this PR if they’re applied. -->
<!-- Please put an "X" in the box on ones that apply. -->
<!-- For more details on what pods are affected by each label, see the wiki -->
<!-- https://github.com/freelawproject/courtlistener/wiki/Pull-requests-%60skip%E2%80%90%7Btype%7D%E2%80%90deploy%60-labels -->

<!-- Check here if the entire deployment can be skipped -->
<!-- This might be the case for a small fix, a tweak to documentation or something like that. -->
- [ ] `skip-deploy` (skips everything below)
    <!-- Check here if the web tier can be skipped -->
    <!-- This is the case if you're working on code that doesn't affect the front end, like management commands, tasks, or documentation. -->
    - [x] `skip-web-deploy`
    <!-- Check here if the deployment to celery can be skipped -->
    <!--This is the case if you make no changes to tasks.py or the code that tasks rely on. -->
    - [ ] `skip-celery-deploy`
    <!-- check this if deployment to cron jobs can be skipped -->
    <!-- This is the case if no changes are made that affect cronjobs. -->
    - [x] `skip-cronjob-deploy`
    <!-- Deployment of daemons can be skipped -->
    <!-- This is the case if you haven't updated daemons or the code they depend on. -->
    - [x] `skip-daemon-deploy`

mlissner reviewed (commented):
This looks interesting, but I've lost track of why we need to do the cache. Can you explain that a bit, and maybe make a comment explaining it, if need be? 

I agree that `html_with_citations` is the right thing to focus on, since it's supposed to always be there. 

I'm also a bit scared to put embeddings into redis. We did that in an earlier task and they took up a lot of space. I forget how Alberto solved that. OTOH, if we wanted to put them into an S3 cache...

This will be nice to have. One of the last pieces!

ERosendo commented:
> Can you explain that a bit, and maybe make a comment explaining it, if need be?

We’re using Redis to pass embeddings through the Celery task chain. When we first introduced the task for creating embeddings in https://github.com/freelawproject/courtlistener/pull/5046, we relied on Celery to pass the response directly between tasks. That approach led to a Redis issue because `CELERY_RESULT_EXPIRES` is set to 60 minutes, causing embeddings to accumulate in memory even after the chain finished.

>  I forget how Alberto solved that.

The fix was to store embeddings in Redis, pass only a reference to downstream tasks, and delete the data from the cache as soon as processing completes. Here's the PR that introduced the fix: https://github.com/freelawproject/courtlistener/pull/5423

> OTOH, if we wanted to put them into an S3 cache...
> 
> This will be nice to have. One of the last pieces! 

That's a great idea.

mlissner commented:
> The fix was to store embeddings in Redis, pass only a reference to downstream tasks, and delete the data from the cache as soon as processing completes. Here's the PR that introduced the fix: https://github.com/freelawproject/courtlistener/pull/5423

Oh. Exactly like you've done here. Got it. :)

albertisfu commented on cl/search/tasks.py:
```diff
@@ -1770,6 +1792,55 @@ def get_embeddings_cache_key(batch_uuid: str, batch_range: str) -> str:
+        )
+    )
+    # Exit early if the microservice call failed
+    if not embeddings.is_success:
```
Sometimes when Inception fails to generate embeddings, the result is not a list but a dictionary containing the response error.

I think it’s a good idea to handle this case so we don’t try to index anything that isn’t actually embeddings, and so we can acknowledge the error for debugging.

We can do something similar to what we did in the save_embeddings task:

    if not isinstance(embeddings, list):
        if isinstance(embeddings, dict):
            logger.error(
                "Received API error response in embeddings: %s",
                json.dumps(embeddings, default=str),
            )
        else:
            logger.error(
                "Unexpected data type for embeddings: %s (%s)",
                str(embeddings)[:200],
                type(embeddings),
            )

ERosendo replied:
Thanks for the suggestion. I'll update the task

albertisfu replied:
Great! I was wondering if we can avoid this duplicated code, which is also present in save_embeddings. Maybe we could create a method to log the errors and reuse it in both places?

albertisfu reviewed (changes requested):
Thanks @ERosendo this looks great.

Just a couple of comments.

I also think it would be a good idea to put embedding generation and indexing behind a Django setting, so we can control when to start this process, considering we’ll need to have an Inception CPU instance running beforehand.

ERosendo commented on cl/search/signals.py:
```diff
@@ -455,6 +455,7 @@
                 "html_anon_2020": ["text"],
                 "html": ["text"],
                 "plain_text": ["text"],
+                "html_with_citations": ["text"],
```
> Is the intention of adding html_with_citations to the Opinion mapping so that it can populate the OpinionDocument.text field in ES?

That’s correct!

> If so, I think we also need to add html_with_citations to the OpinionDocument.prepare_text method.

I completely agree. I actually tried adding it as part of this PR, but several tests started failing. To avoid broadening the scope here, I think it’s best to handle that in a follow-up PR.

albertisfu reviewed (commented):
Thanks @ERosendo just one additional comment.

The usage of the S3 cache will be moved once we have django-s3-express-cache ready to use, right?

ERosendo commented:
Alberto, thanks for the latest comment. I’ve refactored the logic, and this PR is ready for another pass.

I also forgot to mention that I added a new environment variable called `ENABLE_EMBEDDING_COMPUTATION`, which controls when we start creating embeddings for opinions. By default, it’s set to **false**.

> The usage of the S3 cache will be moved once we have django-s3-express-cache ready to use, right?

That's correct
